storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -2812,6 +2816,60 @@ fseg_free_extent(
 	return DB_SUCCESS;
 }
 
+/** Free the extent and fragment page associated with
+the segment.
+@param iblock  page where segment header are placed
+@param inode   index node information
+@param space   tablespace where segment resides
+@param mtr     mini-transaction
+@param header  segment header
+@param all_pages_freed Indicate that all pages are freed
+@param ahi     Adaptive hash index
+@return whether freeing was completed, expect the segment node */
+static
+bool fseg_free_step_low(buf_block_t *iblock, fseg_inode_t *inode,
+                        fil_space_t *space, mtr_t *mtr,
+                        fseg_header_t *header, bool *all_pages_freed
+#ifdef BTR_CUR_HASH_ADAPT
+                        , bool ahi=false
+#endif /* BTR_CUR_HASH_ADAPT */
+                        )
+{
+  dberr_t err= DB_SUCCESS;
+  if (xdes_t* descr= fseg_get_first_extent(inode, space, mtr, &err))
+    /* Free the extent held by the segment */
+    return fseg_free_extent(inode, iblock, space,
+                            xdes_get_offset(descr), mtr
+#ifdef BTR_CUR_HASH_ADAPT
+                            , ahi
+#endif /* BTR_CUR_HASH_ADAPT */
+                           ) != DB_SUCCESS;
+  if (err != DB_SUCCESS)
+    return true;
```

Comment:
The first `return` could be a tail call and the second one would avoid loading a constant if this function returned `dberr_t` instead of `bool`.